src/api_manager/check_security_rules.cc

```diff
@@ -89,24 +98,60 @@ class AuthzChecker : public std::enable_shared_from_this<AuthzChecker> {
 
 AuthzChecker::AuthzChecker(ApiManagerEnvInterface *env,
                            auth::ServiceAccountToken *sa_token)
-    : env_(env), sa_token_(sa_token) {}
+    : env_(env), sa_token_(sa_token){}
 
-void AuthzChecker::Check(
-    std::shared_ptr<context::RequestContext> context,
-    std::function<void(Status status)> final_continuation) {
+
+bool AuthzChecker::CheckCache(std::shared_ptr<context::RequestContext> context,
+                              std::function<void(Status status)> final_continuation) {
   // TODO: Check service config to see if "useSecurityRules" is specified.
   // If so, call Firebase Rules service TestRuleset API.
-
   if (!context->service_context()->IsRulesCheckEnabled() ||
       context->method() == nullptr || !context->method()->auth()) {
     env_->LogInfo("Skipping Firebase Rules checks since it is disabled.");
     final_continuation(Status::OK);
-    return;
+    return false;
   }
 
-  // Fetch the Release attributes and get ruleset name.
-  auto checker = GetPtr();
-  HttpFetch(GetReleaseUrl(*context), kHttpGetMethod, "",
+  bool cache_hit = false;
+  // Cached authorization result, default to false.
+  bool pre_authz_success = false;
+
+  AuthzValue val;
+  if (context->service_context()->authz_cache().Lookup(context->request()->GetAuthToken(),
+                                                   context->request()->GetRequestPath(),
+                                                   context->request()->GetRequestHTTPMethod(),
+                                                   &val)) {
+    // Cache hit happens only if cache entry is not expired.
+    if (system_clock::now() <= val.exp) {
+      cache_hit = true;
+      pre_authz_success = val.if_success;
+    } else {
+      // Remove cache entry if expired.
+      context->service_context()->authz_cache().Delete(context->request()->GetAuthToken(),
+                                                       context->request()->GetRequestPath(),
+                                                       context->request()->GetRequestHTTPMethod());
+    }
+  }
+  if (cache_hit) {
+    GetPtr()->env_->LogDebug("Cache hit.");
+    if (pre_authz_success) {
+      final_continuation(Status::OK);
+    } else {
+      final_continuation(Status(Code::PERMISSION_DENIED, std::string("Unauthorized access (cached).")));
+    }
+    return false;
```

Comment:
this is confusing.   CheckCache() should return true if cache hit,  false for cache miss